app/scripts/onUpdate.ts

```diff
@@ -0,0 +1,120 @@
+  requestSafeReload: () => void,
+): void {
+  const { appStateController } = controller;
+  const { lastUpdatedFromVersion } = appStateController.state;
+  const isFirefox = getPlatform() === PLATFORM_FIREFOX;
+
+  log.debug('[onUpdate]: Update installation detected');
+  log.info(`[onUpdate]: Updated from version ${previousVersion}`);
+  log.info(
+    `[onUpdate]: Recorded last updated from version: ${lastUpdatedFromVersion}`,
+  );
+  log.info(`[onUpdate]: isFirefox: ${isFirefox}`);
+  log.info(`[onUpdate]: Current version: ${platform.getVersion()}`);
+
+  // Browser might trigger an update event even when the version hasn't changed,
+  // like when reloading the extension manually.
+  if (previousVersion === lastUpdatedFromVersion) {
+    return;
+  }
+
+  const lastUpdatedAt = Date.now();
+
+  if (!isFirefox) {
+    // Work around Chromium bug https://issues.chromium.org/issues/40805401
+    // by doing a safe reload after an update. We'll be able to gate this
+    // behind a Chromium version check once we know the chromium version #
+    // that fixes this bug, ETA: December 2025 (likely in `143.0.7465.0`).
+    // Once we no longer support the affected Chromium versions, we should
+    // remove this workaround.
+    // We only want to do the safe reload when the version actually changed,
+    // just as a safe guard, as Chrome fires this event each time we call
+    // `runtime.reload` -- as we really don't want to send Chrome into a restart
+    // loop! This is overkill, as `onUpdate` is already only called when
+    // `previousVersion !== platform.getVersion()`, but better safe than better
+    // safe than better safe than... rebooting forever. :-)
+    const { remoteFeatureFlagController } = controller;
+    const shouldReload = remoteFeatureFlagController.state.remoteFeatureFlags
+      .extensionPlatformAutoReloadAfterUpdate as boolean | undefined;
+    log.info(`[onUpdate]: Should reload: ${shouldReload}`);
+    if (shouldReload === true) {
+      controller.store.on('update', handleUpdate);
+    }
+  }
+  appStateController.setLastUpdatedAt(lastUpdatedAt);
+  appStateController.setLastUpdatedFromVersion(previousVersion);
+
+  /**
+   * We wait for the store to update; when `update` is called the
+   * persistence layer will start writing the state to the database. Once
+   * that starts we will `requestSafeReload` to ensure its the last state
+   * update before the reload happens.
+   *
+   * @param newState - The new MetaMask state.
+   */
+  function handleUpdate(newState: MetaMaskStateType) {
+    log.info('[onUpdate]: Controller store `update` event fired.');
+    // ensure the new state matches the lastUpdatedFromVersion, otherwise
+    // we might reload before the state is fully persisted, potentially
+    // causing a restart loop.
+    const appStateControllerState =
+      newState.AppStateController as AppStateControllerState;
+    if (
+      appStateControllerState.lastUpdatedFromVersion === previousVersion &&
```

Comment:
... I just realized what state we're using. I thought we were going to use `appMetadataController.previousAppVersion`, I didn't realize the same state was also tracked in `appStateController` (why is it duplicated 🤔 )

The `appMetadataController` gets updated automatically upon construction, so that would save us having to wait for the next state update event. We could check it immediately here.

Reply:
Yeah, I was just using what was already there (as you pointed out in the slack thread). Not sure what do to going forward.